### PR TITLE
Relax restrictions for external buffers

### PR DESCRIPTION
**Pull request by Pessimistress**
For #3447 

#### Change List

- Only cast external buffer array type if necessary
- Allow `attribute.type` to be overwritten

### CODE REVIEW

File: test/modules/core/lib/attribute-manager.spec.js [hunk 1]
```diff
@@ -176,7 +176,7 @@ test('AttributeManager.update - external buffers', t => {
     }
   });
 
-  t.is(attribute.buffer.accessor.type, gl.UNSIGNED_BYTE, 'colors casted to correct type');
+  t.is(attribute.type, gl.FLOAT, 'colors accessor is set to correct type');
```
Comment (ibgreen) [inline test/modules/core/lib/attribute-manager.spec.js]: A test case setting positions to UNSIGNED_INT would be perfect, but I gather this comes in next PR.

File: modules/core/src/lib/attribute.js [hunk 2]
```diff
@@ -337,23 +337,35 @@ export default class Attribute extends BaseAttribute {
...
     this.update(opts);
+
     state.needsRedraw = true;
 
-    this._updateShaderAttributes();
     return true;
   }
 
   // PRIVATE HELPER METHODS
+  _checkExternalBuffer(opts) {
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Nit: naming, `check...` makes it sounds like a read only operation.

File: modules/core/src/lib/attribute.js [hunk 3]
```diff
@@ -337,23 +337,35 @@ export default class Attribute extends BaseAttribute {
...
+    if (!opts.constant && opts.value) {
+      const ArrayType = glArrayFromType(this.defaultType);
+      if (
+        opts.value.BYTES_PER_ELEMENT !== ArrayType.BYTES_PER_ELEMENT &&
+        this.hasShaderAttributes
+      ) {
+        // Shader attributes have hard-coded offsets and strides
+        // TODO - switch to element offsets and element strides?
+        log.warn(`Attribute ${this.id} is casted to ${ArrayType.name}`)();
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Nit: "casted" seems a little ambiguous to me in a JS array context (e.g. could be `new ArrayType(opts.value.buffer)`). Maybe just say "copied"

File: modules/core/src/lib/attribute.js [hunk 4]
```diff
@@ -337,23 +337,35 @@ export default class Attribute extends BaseAttribute {
...
     return true;
   }
 
   // PRIVATE HELPER METHODS
+  _checkExternalBuffer(opts) {
+    if (!opts.constant && opts.value) {
+      const ArrayType = glArrayFromType(this.defaultType);
+      if (
+        opts.value.BYTES_PER_ELEMENT !== ArrayType.BYTES_PER_ELEMENT &&
+        this.hasShaderAttributes
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Nit: this.hasShaderAttributes is the primary condition, right, so place it first in the `&&`?

File: modules/core/src/lib/attribute.js [hunk 5]
```diff
@@ -36,6 +36,7 @@ export default class Attribute extends BaseAttribute {
     let {defaultValue = [0, 0, 0, 0]} = opts;
     defaultValue = Array.isArray(defaultValue) ? defaultValue : [defaultValue];
 
+    this.defaultType = this.type || GL.FLOAT;
```
Comment (ibgreen) [inline modules/core/src/lib/attribute.js]: Nit: Feels like this should go into `BaseAttribute` - getting hard to scan here and see that we have both `type` and `defaultType` and grok the difference.

**Comment by Pessimistress** [comment]
> It is a good question if the "casting" for shaderAttributes is worth keeping

I've switched to throwing an error.

**Pull request merged**